ERCS/erc-5791.md

```diff
@@ -129,34 +136,63 @@ This proposal is backward compatible with [ERC-721](./eip-721.md) on an API leve
-  }
-  unchecked {
-    if (block.number - blockNumberUsedInSig > getMaxBlockhashValidWindow()) {
-      revert BlockNumberTooOld();
+/// @dev Transfers the `tokenId` assigned to `chipId` to `to`.
+function transferToken(
+    address to,
+    address chipId,
+    bytes memory chipSignature,
+    uint256 signatureTimestamp,
+    bool useSafeTransfer,
+    bytes memory extras
+) public virtual returns (uint256 tokenId) {
+    tokenId = tokenIdFor(chipId);
+    _validateSigAndUpdateNonce(to, chipId, chipSignature, signatureTimestamp, extras);
+    if (useSafeTransfer) {
+        _safeTransfer(ownerOf(tokenId), to, tokenId, "");
+    } else {
+        _transfer(ownerOf(tokenId), to, tokenId);
+    }
+}
+
+/// @dev Validates the `chipSignature` and update the nonce for the future signature of `chipId`.
+function _validateSigAndUpdateNonce(
+    address to,
+    address chipId,
+    bytes memory chipSignature,
+    uint256 signatureTimestamp,
+    bytes memory extras
+) internal virtual {
+    bytes32 hash = _getSignatureHash(signatureTimestamp, chipId, to, extras);
+    if (!SignatureCheckerLib.isValidSignatureNow(chipId, hash, chipSignature)) {
+        revert InvalidSignature();
     }
-  }
-  bytes32 blockHash = blockhash(blockNumberUsedInSig);
-  bytes32 signedHash = keccak256(abi.encodePacked(_msgSender(), blockHash))
-    .toEthSignedMessageHash();
-  address chipAddr = signedHash.recover(signatureFromChip);
+    chipNonce[chipId] = bytes32(uint256(hash) ^ uint256(blockhash(block.number - 1)));
+}
+
+/// @dev Returns the digest to be signed by the `chipId`.
+function _getSignatureHash(uint256 signatureTimestamp, address chipId, address to, bytes memory extras)
+    internal
+    virtual
+    returns (bytes32)
+{
+    if (signatureTimestamp > block.timestamp) revert SignatureTimestampInFuture();
+    if (signatureTimestamp + maxDurationWindow < block.timestamp) revert SignatureTimestampTooOld();
+    bytes32 hash = keccak256(
+        abi.encode(address(this), block.chainid, chipNonce[chipId], to, signatureTimestamp, keccak256(extras))
+    );
+    return ECDSA.toEthSignedMessageHash(hash);
 }
 
 ```
 
 ## Security Considerations
 
-The [ERC-191](./eip-191.md) signature passed to `transferTokenWithChip` requires the function caller's address in its signed data so that the signature cannot be used in a replay attack. It also requires a recent blockhash so that a malicious chip owner cannot pre-generate signatures to use after a short time window (e.g. after the owner of the physical item changes).
+The [ERC-191](./eip-191.md) signature passed to `transferToken` requires the function caller's address in its signed data so that the signature cannot be used in a replay attack. It also requires a recent block timestamp so that a malicious chip owner cannot pre-generate signatures to use after a short time window (e.g. after the owner of the physical item changes). It's recommended to use a non-deterministic nonce when generating signatures.
```

Comment:
> It's recommended to use a non-deterministic nonce when generating signatures.

This is somewhat ambiguous to me; is this the `k` value used in the actual ECDSA signature? Or the nonce stored in  the contract for `timestamp` purposes? If the latter I might recommend referencing the implementation directly -- e.g. "...use a non-deterministic `chipNonce` when..." or "stored nonce".